src/mod_mam_sql.erl

```diff
@@ -117,42 +118,90 @@ store(Pkt, LServer, {LUser, LHost}, Type, Peer, Nick, _Dir, TS) ->
 		  fxml:element_to_binary(Pkt)
 	  end,
 	case SqlType of
-	  mssql -> case ejabberd_sql:sql_query(
-	           LServer,
-	           ?SQL_INSERT(
-	              "archive",
-	              ["username=%(SUser)s",
-	               "server_host=%(LServer)s",
-	               "timestamp=%(TS)d",
-	               "peer=%(LPeer)s",
-	               "bare_peer=%(BarePeer)s",
-	               "xml=N%(XML)s",
-	               "txt=N%(Body)s",
-	               "kind=%(SType)s",
-	               "nick=%(Nick)s"])) of
-		{updated, _} ->
-		    ok;
-		Err ->
-		    Err
-	    end;
-	    _ -> case ejabberd_sql:sql_query(
-	           LServer,
-	           ?SQL_INSERT(
-	              "archive",
-	              ["username=%(SUser)s",
-	               "server_host=%(LServer)s",
-	               "timestamp=%(TS)d",
-	               "peer=%(LPeer)s",
-	               "bare_peer=%(BarePeer)s",
-	               "xml=%(XML)s",
-	               "txt=%(Body)s",
-	               "kind=%(SType)s",
-	               "nick=%(Nick)s"])) of
-		{updated, _} ->
-		    ok;
-		Err ->
-		    Err
-	    end
+	  mssql ->
+			case Thread of
+				<<"null">> ->
+					case ejabberd_sql:sql_query(
+								LServer,
+								?SQL_INSERT(
+										"archive",
+										["username=%(SUser)s",
+										"server_host=%(LServer)s",
+										"timestamp=%(TS)d",
+										"peer=%(LPeer)s",
+										"bare_peer=%(BarePeer)s",
+										"xml=N%(XML)s",
+										"txt=N%(Body)s",
+										"kind=%(SType)s",
+										"nick=%(Nick)s"])) of
+						{updated, _} ->
+							ok;
+						Err ->
+							Err
+						end;
+				_ ->
+					case ejabberd_sql:sql_query(
+								LServer,
+								?SQL_INSERT(
+										"archive",
+										["username=%(SUser)s",
+										"server_host=%(LServer)s",
+										"timestamp=%(TS)d",
+										"peer=%(LPeer)s",
+										"bare_peer=%(BarePeer)s",
+										"xml=N%(XML)s",
+										"txt=N%(Body)s",
+										"kind=%(SType)s",
+										"nick=%(Nick)s",
+										"thread=%(Thread)s"])) of
+						{updated, _} ->
+							ok;
+						Err ->
+							Err
+						end
+			end;
+		_ ->
+			case Thread of
+				<<"null">> ->
+					case ejabberd_sql:sql_query(
+								LServer,
+								?SQL_INSERT(
+										"archive",
+										["username=%(SUser)s",
+										"server_host=%(LServer)s",
+										"timestamp=%(TS)d",
+										"peer=%(LPeer)s",
+										"bare_peer=%(BarePeer)s",
+										"xml=%(XML)s",
+										"txt=%(Body)s",
+										"kind=%(SType)s",
+										"nick=%(Nick)s"])) of
+						{updated, _} ->
+							ok;
+						Err ->
+							Err
+						end;
+				_ ->
```

Comment:
In all other cases, we insert the value inside the `<thread>` element. We make an assumption that all messages sent through will include `<thread>`, which seems to be true based on the other ejabberd code I've investigated.

Reply:
Note to self - if the <thread> element is not present, this attempts to post the empty string to `thread` instead of `NULL`. Need to update the previous case to be `<<"null">>` or `<<"">>`